core/system/src/page_allocator.c

```diff
@@ -0,0 +1,333 @@
+/*
+ * Copyright (c) 2016, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* This file can be compiled externally to provide the page allocator algorithm
+ * for devices NOT supported by uVisor. For this purpose this file is copied as
+ * is into the target build folder and compiled by the target build system. */
+#if defined(UVISOR_PRESENT) && (UVISOR_PRESENT == 1)
+
+#include <uvisor.h>
+#include "page_allocator.h"
+#include "mpu/vmpu_unpriv_access.h"
+#include "mpu/vmpu.h"
+#include "halt.h"
+#include "context.h"
+#define page_table_read  vmpu_unpriv_uint32_read
+#define page_table_write vmpu_unpriv_uint32_write
+/* For uVisor in supported mode the page allocator is always called through a
+ * SVC call, which automagically serializes access to it. */
+#define UVISOR_PAGE_ALLOCATOR_MUTEX_AQUIRE  {}
+#define UVISOR_PAGE_ALLOCATOR_MUTEX_RELEASE {}
+
+#else /* defined(UVISOR_PRESENT) && (UVISOR_PRESENT == 1) */
+
+#include "uvisor-lib/uvisor-lib.h"
+#define DPRINTF(...) {}
+#define g_active_box 0
+#define vmpu_is_box_id_valid(...) 0
+#define vmpu_public_flash_addr(...) 1
+#define vmpu_sram_addr(...) 1
+/* HALT_ERROR can be provided by the OS. For example using `assert`. */
+#ifndef HALT_ERROR
+#define HALT_ERROR(id, ...) DPRINTF(__VA_ARGS__)
+#endif
+/* Mutex implementation must be provided by the OS, since the page allocator
+ * is called directly from user code instead of through an SVC call. */
+#ifndef UVISOR_PAGE_ALLOCATOR_MUTEX_AQUIRE
+#define UVISOR_PAGE_ALLOCATOR_MUTEX_AQUIRE {}
```

Comment:
Should we `#error` here instead of silently failing to acquire a mutex if the uVisor porting engineer forgot to implement this macro?